### PR TITLE
Adapt `SklearnIntentClassifier` to `GraphComponent` interface

### LINKED ISSUE #9406: Adapt `SklearnIntentClassifier` to new `GraphComponent` interface
**Overview of the Solution**:
Adapt the `SklearnIntentClassifier ` to the new `GraphComponent` interface.
See the [implementation proposal](https://www.notion.so/rasa/Rasa-Open-Source-3-0-Architecture-Implementation-Proposal-51ab90b05c41435ca98189a101676a1e) for further details.

* adapt `create` and `load` functions to new signatures
* implement `required_packages` and `supported_languages`
* adapt persistence
* ~split train and train process~ (this is not necessary for classifiers)
* support multiple messages in `process` (**make sure to handle empty messages**)

**Definition of Done**:
- [x] `SklearnIntentClassifier` adapted to the new GraphComponent interface
- [x] Unit tests are adapted to adapted component
- [x] Other tests (integration tests, etc.) don't break (because they'd (for now) still use the old component implementation)

### PR DESCRIPTION
**Pull request by ancalita**
**Proposed changes**:
- closes #9406 

**Status (please check what you already did)**:
- [X] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [X] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 1]
```diff
@@ -1,72 +1,92 @@
...
         le: Optional["sklearn.preprocessing.LabelEncoder"] = None,
     ) -> None:
         """Construct a new intent classifier using the sklearn framework."""
         from sklearn.preprocessing import LabelEncoder
 
-        super().__init__(component_config)
+        self.component_config = config
+        self._model_storage = model_storage
+        self._resource = resource
+        self._execution_context = execution_context
```
Comment (wochinge) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:68]: do we actually need to save this one?

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 2]
```diff
@@ -83,14 +103,8 @@ def transform_labels_num2str(self, y: np.ndarray) -> np.ndarray:
 
         return self.le.inverse_transform(y)
 
-    def train(
-        self,
-        training_data: TrainingData,
-        config: Optional[RasaNLUModelConfig] = None,
-        **kwargs: Any,
-    ) -> None:
+    def train(self, training_data: TrainingData, **kwargs: Any,) -> Resource:
```
Comment (wochinge) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:106]: let's pass in `num_threads` via the config and drop the `kwargs`. Seems cleaner to me.

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 3]
```diff
@@ -225,40 +244,54 @@ def predict(self, X: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
...
-            return cls(meta, classifier, encoder)
-        else:
-            return cls(meta)
+        try:
+            with model_storage.read_from(resource) as model_dir:
+                file_name = cls.__name__
+                classifier_file = list(model_dir.glob(f"**/{file_name}_classifier.pkl"))
+
+                if classifier_file:
+                    assert len(classifier_file) == 1
```
Comment (wochinge) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:276]: let's not have an `assert` in production code. How about raising an exception which is then caught by the block in 292?

File: tests/nlu/classifiers/test_sklearn_classifier.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+from rasa.nlu import registry
+from rasa.nlu.classifiers.sklearn_intent_classifier import (
+    SklearnIntentClassifierGraphComponent,
+)
+from rasa.nlu.config import RasaNLUModelConfig
+from rasa.shared.nlu.training_data.training_data import TrainingData
+
+
+@pytest.fixture()
+def training_data(nlu_data_path: Text):
```
Comment (wochinge) [inline tests/nlu/classifiers/test_sklearn_classifier.py:19]:
```suggestion
def training_data(nlu_data_path: Text) -> TrainingData:
```

File: tests/nlu/classifiers/test_sklearn_classifier.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+):
+    return SklearnIntentClassifierGraphComponent.create(
+        SklearnIntentClassifierGraphComponent.get_default_config(),
+        default_model_storage,
+        Resource("sklearn"),
+        default_execution_context,
+    )
+
+
+def test_persist_and_load(
```
Comment (wochinge) [inline tests/nlu/classifiers/test_sklearn_classifier.py:35]: can we add a test where loading the component from the model storage fails and we initialize an empty `SklearnClassifier`?

File: rasa/nlu/classifiers/sklearn_intent_classifier.py (hunk 1 above)
Comment (ancalita) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py:68]: True, `execution_context` is not required, I'll remove it; the others are needed.

File: rasa/nlu/classifiers/sklearn_intent_classifier.py (hunk 3 above)
Comment (ancalita) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py:276]: Sure, what would be the exception? 
Here my intention was to make sure that `len` is not greater than 1. Perhaps I can raise a debug warning that more than 1 file were found? In theory it shouldn't be more than 1 file, maybe I'm overcomplicating matters.

Comment (wochinge) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py:276]: how about just throwing a `ValueError`? Your `except ValueError` would then catch it.

Comment (ancalita) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py:276]: Thinking more about this approach - it would return a new component instance when actually files were found that could be used to load the component. Maybe I'm imagining a scenario that actually would never happen (e.g. have more than 1 file found) 🤔 I will just remove the assertion, WDYT?

Comment (ka-bu) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py:276]: were not testing this anywhere else - we're always relying on the assumption that only persist() writes to that directory - so I think it's ok to remove the check for "== 1" :)  ... we could also create some constants for classifier_file_name and encoder_file_name and just use those here and in persist (because we also rely on the model storage/resource to be a unique location for this component evyerwhere)

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 6]
```diff
@@ -165,49 +182,51 @@ def _create_classifier(
...
 
-    def process(self, message: Message, **kwargs: Any) -> None:
+    def process(self, messages: List[Message]) -> List[Message]:
         """Return the most likely intent and its probability for a message."""
+        for message in messages:
+            if not self.clf:
+                # component is either not trained or didn't
+                # receive enough training data
+                intent = None
+                intent_ranking = []
```
Comment (wochinge) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:192]: should we pull this out of the loop?

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 7]
```diff
@@ -225,40 +244,53 @@ def predict(self, X: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
...
-            classes = io_utils.json_unpickle(encoder_file)
-            encoder = LabelEncoder()
-            encoder.classes_ = classes
-            return cls(meta, classifier, encoder)
-        else:
-            return cls(meta)
+        try:
+            with model_storage.read_from(resource) as model_dir:
+                file_name = cls.__name__
+                classifier_file = list(model_dir.glob(f"**/{file_name}_classifier.pkl"))
```
Comment (wochinge) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:273]: why do we actually need to glob this at all? Don't we exactly know where to load it from as we explicitly hardcode the file names?

File: tests/nlu/classifiers/test_sklearn_classifier.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    default_sklearn_intent_classifier.train(training_data)
+
+    loaded = SklearnIntentClassifierGraphComponent.load(
+        SklearnIntentClassifierGraphComponent.get_default_config(),
+        default_model_storage,
+        Resource("sklearn"),
+        default_execution_context,
+    )
+
+    predicted = copy.copy(training_data)
```
Comment (wochinge) [inline tests/nlu/classifiers/test_sklearn_classifier.py:66]: I think a `copy` is not enough - you'll need a `deepcopy`

File: rasa/nlu/classifiers/sklearn_intent_classifier.py (hunk 6 above)
Comment (ancalita) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py:192]: And then return the list with unmodified messages?

File: rasa/nlu/classifiers/sklearn_intent_classifier.py (hunk 7 above)
Comment (ancalita) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py:273]: Oh right - would it be enough to just replace this with `model_dir / f"{file_name}_classifier.pkl"` ?

File: rasa/nlu/classifiers/sklearn_intent_classifier.py (hunk 6 above)
Comment (wochinge) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py:192]: mhm, actually that makes sense 👍🏻 Let's keep it in then

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 9]
```diff
@@ -1,72 +1,92 @@
...
+            "num_threads": 1,
+        }
 
     def __init__(
         self,
-        component_config: Optional[Dict[Text, Any]] = None,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
```
Comment (wochinge) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:59]:
```suggestion
```

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 10]
```diff
@@ -225,40 +244,53 @@ def predict(self, X: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
...
-            encoder.classes_ = classes
-            return cls(meta, classifier, encoder)
-        else:
-            return cls(meta)
+        try:
+            with model_storage.read_from(resource) as model_dir:
+                file_name = cls.__name__
+                classifier_file = model_dir / f"{file_name}_classifier.pkl"
+
+                if classifier_file:
```
Comment (wochinge) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:275]:
```suggestion
                if classifier_file.exists():
```

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 11]
```diff
@@ -225,40 +244,53 @@ def predict(self, X: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
...
+                        config,
+                        model_storage,
+                        resource,
+                        execution_context,
+                        classifier,
+                        encoder,
+                    )
+        except ValueError:
+            logger.warning(
+                f"Failed to load {cls.__name__} from model storage. Resource "
```
Comment (wochinge) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:293]:
```suggestion
                f"Failed to load '{cls.__name__}' from model storage. Resource "
```

**Review by wochinge — APPROVED** [review]
Left a few more comment but looks great otherwise 🚀
